packages/contracts/contracts/SmartInvoiceFactoryV2.sol

```diff
@@ -0,0 +1,312 @@
+// SPDX-License-Identifier: MIT
+
+pragma solidity ^0.8.0;
+pragma experimental ABIEncoderV2;
+
+import "@openzeppelin/contracts/proxy/Clones.sol";
+import "@openzeppelin/contracts/access/AccessControl.sol";
+import "./interfaces/ISmartInvoiceFactoryV2.sol";
+import "./interfaces/ISmartInvoiceV2.sol";
+import "hardhat/console.sol";
+
+contract SmartInvoiceFactoryV2 is ISmartInvoiceFactoryV2, AccessControl {
+    uint256 public invoiceCount = 0;
+    mapping(uint256 => address) internal _invoices;
+
+    mapping(address => uint256) public resolutionRates;
+
+    bytes32 public constant ADMIN = keccak256("ADMIN");
+
+    /** @dev marks a deployed contract as a suitable implementation for additional escrow invoices formats */
+    /** @dev mapping(implementationType => mapping(implementationVersion => address)) */
+    mapping(bytes32 => mapping(uint256 => address)) public implementations;
+    mapping(bytes32 => uint256) public implementationsVersions;
+    mapping(address => bool) internal implementationExists;
+
+    event LogNewInvoice(
+        uint256 indexed index,
+        address invoice,
+        uint256[] amounts
+    );
+    event UpdateResolutionRate(
+        address indexed resolver,
+        uint256 indexed resolutionRate,
+        bytes32 details
+    );
+
+    address public immutable wrappedNativeToken;
+
+    constructor(address _wrappedNativeToken) {
+        require(
+            _wrappedNativeToken != address(0),
+            "invalid wrappedNativeToken"
+        );
+        wrappedNativeToken = _wrappedNativeToken;
+
+        _grantRole(DEFAULT_ADMIN_ROLE, msg.sender);
+        _grantRole(ADMIN, msg.sender);
+    }
+
+    function _init(
+        address _invoiceAddress,
+        address _client,
+        address _provider,
+        uint8 _resolverType,
+        address _resolver,
+        uint256[] calldata _amounts,
+        bytes calldata _implementationData,
+        bytes32 _implementationType,
+        uint256 _implementationVersion,
+        address _implementationAddress
+    ) internal {
+        uint256 resolutionRate = resolutionRates[_resolver];
+        if (resolutionRate == 0) {
+            resolutionRate = 20;
+        }
+
+        ISmartInvoiceV2(_invoiceAddress).init(
+            _client,
+            _provider,
+            _encodeResolutionData(_resolverType, _resolver, resolutionRate),
+            _amounts,
+            wrappedNativeToken,
+            _implementationData,
+            _encodeImplementationData(
+                _implementationType,
+                _implementationVersion,
+                _implementationAddress,
+                invoiceCount
+            )
```

Comment:
I don't think the implementation stuff is necessary to pass through as a params here. The type and version is only required to get the current contract address to clone